### PR TITLE
Fix #6497: Support global flags passed in after subcommands

### LINKED ISSUE #6497: Support arbitrary order for CLI flags
We currently distinguish between two types of flags:
- those which precede the subcommand ("global"), e.g. `--no-use-color`
- those which follow the subcommand, and may have different behavior depending on the subcommand, e.g. `--select`, `--full-refresh`

IMO this is pretty confusing for end users! We should really seek to support flags in any order (and resolve any resulting ambiguities by actually renaming flags). I know we implemented the new CLI modeling in accordance with the old CLI model — parity is a good goal to shoot for with any refactor — but it would be great to just support both of these, seamlessly and equivalently:
```
$ dbt --no-use-color run --select one_model
$ dbt run --select one_model --no-use-color
```

### PR DESCRIPTION
**Pull request by aranke**
resolves #6497 

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [ ] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: core/dbt/cli/main.py [hunk 1]
```diff
@@ -118,6 +119,44 @@ def invoke(self, args: List[str], **kwargs) -> dbtRunnerResult:
             )
 
 
+# approach from https://github.com/pallets/click/issues/108#issuecomment-280489786
+def global_flags(func):
```
Comment (jtcohen6) [inline core/dbt/cli/main.py:123]: This is neat!

**Review by jtcohen6 — COMMENTED** [review]
Mea culpa - I had missed adding the 1.5 + 1.6 backport labels to https://github.com/dbt-labs/dbt-core/issues/6497. Ideally we could offer the same consistency for all our CLIs starting with v1.5+. Given how "clean" this change looks, how would you feel about backporting the change (at least the approach if not the exact commit diff) to `1.5.latest` + `1.6.latest`?

**Comment by jtcohen6** [comment]
@aranke That's odd - it's only those three flags which are being difficult? Do you have any sense of why?

**Comment by jtcohen6** [comment]
@aranke Thanks for investigating & opening the ticket! We should try to get that tech debt ticket resolved — it's important for our ability to present consistent documentation & CLI behavior (https://github.com/dbt-labs/docs.getdbt.com/issues/4041) — but I agree that it shouldn't block merging the brunt of the work in this PR.

**Review by gshank — APPROVED** [review]

**Review by emmyoop — APPROVED** [review]
LGTM

**Review by ChenyuLInx — APPROVED** [review]
Thanks!

**Pull request merged**